src/control/security/auth/auth_sys.go

```diff
@@ -66,239 +70,83 @@ func VerifyToken(key crypto.PublicKey, token *Token, sig []byte) error {
+	validAuthFlavors := make([]uint32, len(authStrings))
+	for i := 0; i < len(authStrings); i++ {
+		flavor, ok := Flavor_value["AUTH_" + authStrings[i]]
+		if !ok {
+			return nil, errors.Errorf("auth string %s is not recognized", authStrings[i])
+		}
+		validAuthFlavors[i] = uint32(flavor)
 	}
 
-	for i, g := range groupNames {
-		groupNames[i] = sysNameToPrincipalName(g)
-	}
-	return groupNames, nil
+	return validAuthFlavors, nil
 }
 
-// WithUserAndGroup provides an override to set the user, group, and optional list
-// of group names to be used for the request.
-func (r *CredentialRequest) WithUserAndGroup(userStr, groupStr string, groupStrs ...string) {
-	r.getUser = func(id string) (*user.User, error) {
-		return &user.User{
-			Uid:      id,
-			Gid:      id,
-			Username: userStr,
-		}, nil
-	}
-	r.getGroup = func(id string) (*user.Group, error) {
-		return &user.Group{
-			Gid:  id,
-			Name: groupStr,
-		}, nil
-	}
-	r.getGroupNames = func(*CredentialRequest) ([]string, error) {
-		return groupStrs, nil
+func generateAuthMap() AuthMap {
+	var am = make(AuthMap)
+	for _, req := range CredentialRequests {
+		flavor := req.GetAuthFlavor()
+		_, found := am[flavor]
+		if found {
+			panic(fmt.Errorf("multiple authentication methods with the flavor `%s` were found - confirm that all implementations of `GetAuthName` are unique", Flavor_name[int32(flavor)]))
+		}
+		am[flavor] = req
 	}
+	return am
 }
 
-// GetSignedCredential returns a credential based on the provided domain info and
-// signing key.
-func GetSignedCredential(ctx context.Context, req *CredentialRequest) (*Credential, error) {
-	if req == nil {
-		return nil, errors.Errorf("%T is nil", req)
-	}
-
-	if req.DomainInfo == nil {
-		return nil, errors.New("No domain info supplied")
-	}
-
-	hostname, err := req.hostname()
-	if err != nil {
-		return nil, err
-	}
-
-	userPrinc, err := req.userPrincipal()
-	if err != nil {
-		return nil, err
-	}
-
-	groupPrinc, err := req.groupPrincipal()
-	if err != nil {
-		return nil, err
-	}
-
-	groupPrincs, err := req.groupPrincipals()
-	if err != nil {
-		return nil, err
-	}
-
-	// Craft AuthToken
-	sys := Sys{
-		Stamp:       0,
-		Machinename: hostname,
-		User:        userPrinc,
-		Group:       groupPrinc,
-		Groups:      groupPrincs,
-		Secctx:      req.DomainInfo.Ctx()}
-
-	// Marshal our AuthSys token into a byte array
-	tokenBytes, err := proto.Marshal(&sys)
-	if err != nil {
-		return nil, errors.Wrap(err, "Unable to marshal AuthSys token")
-	}
-	token := Token{
-		Flavor: Flavor_AUTH_SYS,
-		Data:   tokenBytes}
-
-	verifier, err := VerifierFromToken(req.SigningKey, &token)
-	if err != nil {
-		return nil, errors.WithMessage(err, "Unable to generate verifier")
+type (
+	AuthMap map[Flavor]CredentialRequestFactory
+
+	CredentialRequestFactory interface {
+		// Allocate, and return, an instance of the type that implements CredentialRequest interface.
+		AllocCredentialRequest() CredentialRequest
```

Comment:
See if factory can do the initialization! Seems redundant to do alloc then init.